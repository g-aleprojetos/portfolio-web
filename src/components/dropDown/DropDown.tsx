import React, {useState, useEffect, useMemo} from 'react';
import {useTranslation} from 'react-i18next';
import {languages, namespaces} from 'utils/i18n/i18n.constants';
import {useBackgroundContext} from 'context/background';
import FlagBrasilSVG from 'assets/icons/iFlag-Brasil-active.svg';
import FlagEUASVG from 'assets/icons/iFlag-USA-active.svg';
import FlagFranceSVG from 'assets/icons/iFlag-France-active.svg';
import * as S from './DropDown.styles';

interface IOption {
  testId: string;
  icon: string;
  idioma: () => string;
  alt: () => string;
  abreviatura: string;
}

export const Dropdown: React.FC = () => {
  const {themeDark} = useBackgroundContext();
  const {t, i18n} = useTranslation(namespaces.components.dropdown);
  const [isOpen, setIsOpen] = useState(false);
  const [selectedOption, setSelectedOption] = useState<IOption | null>(null);
  const [activeOption, setActiveOption] = useState<IOption | null>(null);

  const options = useMemo(
    () => [
      {
        testId: 'idioma-portugues',
        abreviatura: languages.pt,
        icon: FlagBrasilSVG,
        idioma: () => t('portuguese'),
        alt: () => t('altFlagBrazil'),
      },
      {
        testId: 'idioma-ingles',
        abreviatura: languages.en,
        icon: FlagEUASVG,
        idioma: () => t('english'),
        alt: () => t('altFlagEUA'),
      },
      {
        testId: 'idioma-frances',
        abreviatura: languages.fr,
        icon: FlagFranceSVG,
        idioma: () => t('french'),
        alt: () => t('altFlagFrance'),
      },
    ],
    [t],
  );

  useEffect(() => {
    const resolvedLanguage = i18n.resolvedLanguage;

    const initialOption = options.find(option => {
      if (resolvedLanguage === 'pt')
        return option.testId === 'idioma-portugues';
      if (resolvedLanguage === 'en') return option.testId === 'idioma-ingles';
      if (resolvedLanguage === 'fr') return option.testId === 'idioma-frances';
      return null;
    });

    if (initialOption) {
      setSelectedOption(initialOption);
      setActiveOption(initialOption);
    }
  }, [i18n.resolvedLanguage, options]);

  const handleSelectClick = () => {
    setIsOpen(!isOpen);
  };

  const handleOptionClick = (option: IOption) => {
    setSelectedOption(option);
    setActiveOption(option);
    i18n.changeLanguage(option.abreviatura);
    setIsOpen(false);
  };

  return (
    <S.DropdownContainer>
      <S.Select
        data-testid="dropdown-select"
        $themeDark={themeDark}
        $clicked={isOpen}
        onClick={handleSelectClick}>
        <S.Selected>
          {selectedOption && (
            <S.Flag
              src={selectedOption.icon}
              alt={selectedOption.alt()}
              data-testid={selectedOption.testId}
            />
          )}
          <S.Texto $themeDark={!themeDark}>
            {selectedOption ? selectedOption.idioma() : t('select')}
          </S.Texto>
        </S.Selected>
        <S.Caret $themeDark={themeDark} $rotate={isOpen} />
      </S.Select>
      <S.Menu data-testid="dropdown-menu" $themeDark={themeDark} $open={isOpen}>
        {options.map(option => (
          <S.MenuItem
            key={option.testId}
            $active={activeOption === option}
            $themeDark={themeDark}
            onClick={() => handleOptionClick(option)}
            data-testid={option.testId}>
            <S.ContainerOption>
              <S.Flag src={option.icon} alt={option.alt()} />
              <S.Texto $themeDark={!themeDark}>{option.idioma()}</S.Texto>
            </S.ContainerOption>
          </S.MenuItem>
        ))}
      </S.Menu>
    </S.DropdownContainer>
  );
};
