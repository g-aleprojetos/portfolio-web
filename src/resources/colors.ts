export const colors = {
  white: '#ffffff',
  lightgray: '#d3d3d3',
  black: '#000000',
  black50per: '#00000080',
  darkgray: '#23242a',
  yellow: '#FFE879',
  blue: '#15154E',
  lightblue: '#2196F3',
};
