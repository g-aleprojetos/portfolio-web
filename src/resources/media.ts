const devices = {
  xmobile: 375,
  mobile: 425,
  tablet: 768,
  laptop: 1023,
  xlaptop: 1440,
};

export const media = {
  xmobile: `(max-width: ${devices.xmobile}px)`,
  mobile: `(max-width: ${devices.mobile}px)`,
  tablet: `(max-width: ${devices.tablet}px)`,
  laptop: `(max-width: ${devices.laptop}px)`,
  xlaptop: `(max-width: ${devices.xlaptop}px)`,
};
