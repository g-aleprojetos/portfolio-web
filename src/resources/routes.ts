export default {
  Home: '/',
  Projects: '/projects',
  About: '/about',
  Settings: '/settings',
};
