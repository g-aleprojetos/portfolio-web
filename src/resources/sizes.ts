export default {
  px1: '0.063rem',
  px2: '0.125rem',
  px3: '0.188rem',
  px5: '0.313rem',
  px8: '0.5rem',
  px13: '0.813rem',
  px21: '1.313rem',
  px34: '2.125rem',
  px55: '3.438rem',
  px89: '5.563rem',
  px144: '9rem',
  px233: '14.563rem',
  px377: '23.563rem',
  px610: '38.125rem',
  media_768: '768px',
};
