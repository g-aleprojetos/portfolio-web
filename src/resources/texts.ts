const tamanho = {
  xxsmall: 12,
  xsmall: 14,
  small: 16,
  medium: 18,
  big: 24,
};

const altura = {
  xxsmall: 18,
  xsmall: 22,
  small: 24,
  medium: 26,
  big: 34,
};

export default {tamanho, altura};
