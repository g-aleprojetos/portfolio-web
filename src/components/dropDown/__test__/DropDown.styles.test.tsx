import React from 'react';
import {render} from '@testing-library/react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import {colors} from 'resources/colors';
import * as S from '../DropDown.styles';

describe('Image.styles', () => {
  test('Caret DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.Caret />);
    expect(container).toMatchSnapshot();
  });

  test('Caret DEVE girar ao receber a propriedade $rotate', () => {
    const {container} = render(<S.Caret $rotate={true} />);
    expect(container.firstChild).toHaveStyleRule('transform', 'rotate(180deg)');
  });

  test('Caret DEVE mudar a cor da borda superior ao receber a propriedade $themeDark', () => {
    const {container} = render(<S.Caret $themeDark={true} />);
    expect(container.firstChild).toHaveStyleRule(
      'border-top',
      `8px solid ${colors.black}`,
    );
  });

  test('ContainerOption DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.ContainerOption />);
    expect(container).toMatchSnapshot();
  });

  test('DropdownContainer DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.DropdownContainer />);
    expect(container).toMatchSnapshot();
  });

  test('Flag DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.Flag src="icon.svg" />);
    expect(container).toMatchSnapshot();
  });

  test('Menu DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.Menu $open={true} />);
    expect(container).toMatchSnapshot();
  });

  test('Menu DEVE ser visível ao receber a propriedade $open', () => {
    const {container} = render(<S.Menu $open={true} />);
    expect(container.firstChild).toHaveStyleRule('opacity', '1');
    expect(container.firstChild).toHaveStyleRule('display', 'block');
  });

  test('Menu DEVE ter o background correto ao receber a propriedade $themeDark como true', () => {
    const {container} = render(<S.Menu $themeDark={true} />);
    expect(container.firstChild).toHaveStyleRule(
      'background',
      colors.background04,
    );
  });

  test('Menu DEVE ter o background correto ao receber a propriedade $themeDark como false', () => {
    const {container} = render(<S.Menu $themeDark={false} />);
    expect(container.firstChild).toHaveStyleRule('background', colors.gunmetal);
  });

  test('Menu DEVE ter a borda correta ao receber a propriedade $themeDark como true', () => {
    const {container} = render(<S.Menu $themeDark={true} />);
    expect(container.firstChild).toHaveStyleRule(
      'border',
      `1px solid ${colors.cadetGrey}`,
    );
  });

  test('Menu DEVE ter a borda correta ao receber a propriedade $themeDark como false', () => {
    const {container} = render(<S.Menu $themeDark={false} />);
    expect(container.firstChild).toHaveStyleRule(
      'border',
      `1px solid ${colors.charcoal}`,
    );
  });

  test('MenuItem DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.MenuItem />);
    expect(container).toMatchSnapshot();
  });

  test('MenuItem DEVE ter background específico ao receber as propriedades $active e $themeDark', () => {
    const {container} = render(<S.MenuItem $active={true} $themeDark={true} />);
    expect(container.firstChild).toHaveStyleRule(
      'background',
      `${colors.midGray}`,
    );
  });

  test('Select DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.Select />);
    expect(container).toMatchSnapshot();
  });

  test('Select DEVE ter sombra específica ao receber a propriedade $clicked', () => {
    const {container} = render(<S.Select $clicked={true} />);
    expect(container.firstChild).toHaveStyleRule(
      'box-shadow',
      `0 0 12px ${colors.lividIndigo}`,
    );
  });

  test('Select DEVE ter o background e borda corretos ao receber a propriedade $themeDark', () => {
    const {container} = render(<S.Select $themeDark={true} $clicked={true} />);
    expect(container.firstChild).toHaveStyleRule(
      'background',
      `${colors.background02}`,
    );
    expect(container.firstChild).toHaveStyleRule(
      'border',
      `2px solid ${colors.background02}`,
    );
  });

  test('Select DEVE ter a borda correta ao receber as propriedades $themeDark como true e $clicked como true', () => {
    const {container} = render(<S.Select $themeDark={true} $clicked={true} />);
    expect(container.firstChild).toHaveStyleRule(
      'border',
      `2px solid ${colors.background02}`,
    );
  });

  test('Select DEVE ter a borda correta ao receber as propriedades $themeDark como true e $clicked como false', () => {
    const {container} = render(<S.Select $themeDark={true} $clicked={false} />);
    expect(container.firstChild).toHaveStyleRule(
      'border',
      `2px solid ${colors.cadetGrey}`,
    );
  });

  test('Select DEVE ter a borda correta ao receber as propriedades $themeDark como false e $clicked como true', () => {
    const {container} = render(<S.Select $themeDark={false} $clicked={true} />);
    expect(container.firstChild).toHaveStyleRule(
      'border',
      `2px solid ${colors.lividIndigo}`,
    );
  });

  test('Select DEVE ter a borda correta ao receber as propriedades $themeDark como false e $clicked como false', () => {
    const {container} = render(
      <S.Select $themeDark={false} $clicked={false} />,
    );
    expect(container.firstChild).toHaveStyleRule(
      'border',
      `2px solid ${colors.charcoalBlue}`,
    );
  });

  test('Selected DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.Selected />);
    expect(container).toMatchSnapshot();
  });

  test('Texto DEVE ser igual ao snapshot', () => {
    const {container} = render(<S.Texto />);
    expect(container).toMatchSnapshot();
  });

  test('Texto DEVE ter a cor correta ao receber a propriedade $themeDark', () => {
    const {container} = render(<S.Texto $themeDark={true} />);
    expect(container.firstChild).toHaveStyleRule(
      'color',
      `${colors.background01}`,
    );
  });

  test('MenuItem DEVE mudar a cor ao receber hover e a propriedade $themeDark', () => {
    const {container} = render(<S.MenuItem $themeDark={true} />);
    expect(container.firstChild).toHaveStyleRule(
      'background',
      `${colors.darkerCadetGrey}`,
      {
        modifier: ':hover',
      },
    );
  });
});
