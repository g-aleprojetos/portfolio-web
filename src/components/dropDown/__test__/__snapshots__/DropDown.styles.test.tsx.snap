// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image.styles Caret DEVE ser igual ao snapshot 1`] = `
.c0 {
  width: 0;
  height: 0;
  border-left: 4px solid transparent;
  border-right: 4px solid transparent;
  border-top: 8px solid #ffffff;
  transition: transform 0.3s;
  transform: rotate(0deg);
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`Image.styles ContainerOption DEVE ser igual ao snapshot 1`] = `
.c0 {
  display: flex;
  align-items: center;
  gap: 8px;
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`Image.styles DropdownContainer DEVE ser igual ao snapshot 1`] = `
.c0 {
  min-width: 150px;
  position: relative;
  margin: 8px;
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`Image.styles Flag DEVE ser igual ao snapshot 1`] = `
.c0 {
  display: flex;
  width: 24px;
  height: 24px;
  box-sizing: border-box;
}

.c1 {
  pointer-events: auto;
  cursor: default;
}

<div>
  <div
    class="c0 "
    data-testid="container-image"
  >
    <svg
      class="c1"
      data-testid="svg"
      src="icon.svg"
      style="transform: rotate(0deg);"
    />
  </div>
</div>
`;

exports[`Image.styles Menu DEVE ser igual ao snapshot 1`] = `
.c0 {
  position: absolute;
  list-style: none;
  background: #323741;
  border: 1px solid #363a43;
  box-shadow: 0 8px 16px rgba(0, 0, 0, 0.2);
  border-radius: 8px;
  color: #9fa5b5;
  top: 64px;
  left: 50%;
  width: 100%;
  transform: translateX(-50%);
  opacity: 1;
  transition: opacity 0.2s;
  display: block;
  z-index: 1;
}

<div>
  <ul
    class="c0"
  />
</div>
`;

exports[`Image.styles MenuItem DEVE ser igual ao snapshot 1`] = `
.c0 {
  display: flex;
  padding: 12px 8px;
  margin: 4px;
  border-radius: 8px;
  cursor: pointer;
  background: transparent;
}

.c0:hover {
  background: #2a2d35;
}

<div>
  <li
    class="c0"
  />
</div>
`;

exports[`Image.styles Select DEVE ser igual ao snapshot 1`] = `
.c0 {
  background: #23242a;
  display: flex;
  justify-content: space-between;
  align-items: center;
  border: 2px solid #2a2f3b;
  border-radius: 8px;
  padding: 8px 12px;
  cursor: pointer;
  transition: background 0.3s,box-shadow 0.3s;
  box-shadow: none;
}

.c0:hover {
  background: #323741;
}

<div>
  <div
    class="c0"
  />
</div>
`;

exports[`Image.styles Selected DEVE ser igual ao snapshot 1`] = `
.c0 {
  display: flex;
  align-items: center;
  gap: 8px;
}

<div>
  <span
    class="c0"
  />
</div>
`;

exports[`Image.styles Texto DEVE ser igual ao snapshot 1`] = `
.c0 {
  font-size: 12px;
  font-family: AvertaStd-Regular;
  color: #000000;
  margin-top: 0px;
  text-transform: none;
  cursor: pointer;
}

<div>
  <p
    class="c0 "
    color="#000000"
    cursor="pointer"
  />
</div>
`;
